fendermint/vm/interpreter/src/fvm/exec.rs

```diff
@@ -212,6 +212,36 @@ where
             PowerUpdates::default()
         };
 
+        // check for upgrades in the upgrade_scheduler
+        let chain_id = state.chain_id();
+        let block_height: u64 = state.block_height().try_into().unwrap();
+        if let Some(upgrade) = self.upgrade_scheduler.get(chain_id, block_height) {
+            // TODO: consider using an explicit tracing enum for upgrades
+            tracing::info!(?chain_id, height = block_height, "Executing an upgrade");
+
+            // there is an upgrade scheduled for this height, lets run the migration
+            let res = upgrade.execute(&mut state);
+
+            match res {
+                Ok(new_app_version) => {
+                    tracing::info!(result =? res, "upgrade done");
+
+                    // update the app_version in case there was an upgrade defined at this height which upgraded it
+                    if new_app_version != 0 {
+                        state.update_app_version(|app_version| {
+                            *app_version = new_app_version;
+                        });
+
+                        tracing::info!(app_version = state.app_version(), "upgraded app version");
+                    }
+                }
+                Err(err) => {
+                    tracing::error!(?err, "Upgrade failed");
+                    return Err(err);
+                }
```

Comment:
I reckon you could use `let app_version = upgrade.execute(&mut state).context("upgrade failed")?`; and let the top handler do the logging.